### PR TITLE
Adding icon color choice for workspace icons

### PR DESCRIPTION
**Pull request by ab5tract**
This replaces PR #584 as a single commit against the new TypeScript code. Fixes bug #521.

### CODE REVIEW

File: src/layout/verticalPanel/workspaceList.ts [hunk 1]
```diff
@@ -598,11 +600,22 @@ export class WorkspaceButtonIcon extends St.Widget {
...
             this.buildIcons();
         });
     }
 
+    desaturateIcons() {
+        if (! Me.msThemeManager.panelIconColor) {
+            this.add_effect(this.desaturateEffect);
+        } else {
+            this.remove_effect(this.desaturateEffect);
+        }
```
Comment (PapyElGringo) [inline src/layout/verticalPanel/workspaceList.ts:616]:
```suggestion
        const shouldDesaturate = !Me.msThemeManager.panelIconColor;
        const isDesaturate = this.desaturateEffect !== undefined && this.has_effect(this.desaturateEffect);
        if( shouldDesaturate === isDesaturate) return;
        if (shouldDesaturate) {
            this.desaturateEffect = new Clutter.DesaturateEffect();
            this.add_effect(this.desaturateEffect);
        } else {
            this.remove_effect(this.desaturateEffect);
            delete this.desaturateEffect;
        }
```

File: src/layout/verticalPanel/workspaceList.ts [hunk 2]
```diff
@@ -581,12 +581,14 @@ export class WorkspaceButtonIcon extends St.Widget {
 
     msWorkspace: MsWorkspace;
     appIconList: St.Icon[];
+    desaturateEffect: Clutter.DesaturateEffect;
```
Comment (PapyElGringo) [inline src/layout/verticalPanel/workspaceList.ts:584]: If you take my suggestion we need to add `undefined` as valid type

**Review by PapyElGringo — CHANGES_REQUESTED** [review]

**Comment by ab5tract** [comment]
I did not find a `has_effect` method in the `cluttermm` API docs, however it was possible to do accomplish the same task with `add_effect_with_name` and `get_effect`. For some reason `remove_effect` would not compile when using a name, but it works fine when passed the reference anyway.

**Comment by ab5tract** [comment]
> I did not find a `has_effect` method in the `cluttermm` API docs, however it was possible to accomplish the same task with `add_effect_with_name` and `get_effect`. For some reason `remove_effect` would not compile when using a name, but it works fine when passed the reference anyway.

It's also possible to use the `has_effects` method, though it is less precise since it does not check specifically for this desaturation effect. Easy to change to that though, if you would prefer.

File: src/layout/verticalPanel/workspaceList.ts [hunk 3]
```diff
@@ -581,12 +581,14 @@ export class WorkspaceButtonIcon extends St.Widget {
...
     msWorkspace: MsWorkspace;
     appIconList: St.Icon[];
+    desaturateEffect: Clutter.DesaturateEffect | undefined;
 
     constructor(msWorkspace: MsWorkspace) {
         super();
         this.msWorkspace = msWorkspace;
         this.appIconList = [];
-        this.add_effect(new Clutter.DesaturateEffect());
+        this.desaturateEffect = new Clutter.DesaturateEffect();
```
Comment (PapyElGringo) [inline src/layout/verticalPanel/workspaceList.ts:590]: This line should be deleted no?

**Review by PapyElGringo — CHANGES_REQUESTED** [review]

File: src/layout/verticalPanel/workspaceList.ts (hunk 3 above)
Comment (ab5tract) [reply src/layout/verticalPanel/workspaceList.ts]: Absolutely correct. Good catch, sorry about that!

**Comment by PapyElGringo** [comment]
Thanks !

**Pull request merged**